tests/test_kubernetes_master_actions.py

```diff
@@ -0,0 +1,77 @@
+import base64
+import json
+import pytest
+import sys
+from unittest import mock
+
+
+charms = mock.MagicMock()
+sys.modules['charms'] = charms
+sys.modules['charms.layer'] = charms.layer
+
+from actions import user_actions  # noqa: E402
+
+
+def test_protect_resources():
+    """Verify exception when acting on protected resources."""
+    with pytest.raises(SystemExit):
+        user_actions.protect_resources('admin')
+    with pytest.raises(SystemExit):
+        user_actions.protect_resources('kubelet-X')
+
+
+def test_user_list():
+    """Verify user data is parsed correctly from our secrets."""
+    user = 'admin'
+    secret_id = '{}-secret'.format(user)
+
+    test_data = {
+        'items': [{
+            'metadata': {
+                'name': secret_id,
+            },
+            'data': {
+                'username': base64.b64encode(user.encode('utf-8')).decode('utf-8'),
+            }
+        }]
+    }
+    secrets = json.dumps(test_data).encode('utf-8')
+
+    # we expect a {username: secret_id} dict
+    with mock.patch('actions.user_actions.layer.kubernetes_common.kubectl',
+                    return_value=secrets):
+        secret_data = user_actions.user_list()
+        assert user in secret_data.keys()
+        assert secret_id in secret_data.values()
+
+
+@mock.patch('actions.user_actions.action_get')
+def test_user_create(mock_get):
+    """Verify failure if we create a user that already exists."""
+    user = 'testuser'
+    secret_id = '{}-secret'.format(user)
+    test_data = {
+        user: secret_id
+    }
+
+    mock_get.return_value = user
+    with mock.patch('actions.user_actions.user_list',
+                    return_value=test_data):
+        user_actions.user_create()
+        assert user_actions.action_fail.called
```

Comment:
Correct, that's because `lib.k8s-master.create_secret` is already unit tested in the k8s-master lib unit tests.  That said, there is a difference between the charm and the action here; this action calls `k8s_master.create_secret` *as well as* `k8s_common.create_kubeconfig`, but i don't think this action should unit test those funcs.

Feels better to me for those tests to happen in their respective modules.  LMK if i'm missing a scenario other than "creating an existing user == failure" that should be tested here.

Reply:
My issue with that is that almost the entire function is below that `user in users` check so doesn't get covered. 
 `k8s_common.create_kubeconfig` is already mocked out, and it should be easy to temporarily mock out `layer.k8s_master` so that you can cover the rest of that function without it doing any external function calls.